Calypso build: Use @wordpress webpack plugin

sirreal opened the pull request:
We should keep an eye on consumers getting the same packages:

https://github.com/Automattic/wp-calypso/blob/e85241257c09ba9838850b834923f06f393dc283/packages/wordpress-external-dependencies-plugin/index.js#L20-L38

Compare with https://github.com/WordPress/gutenberg/blob/master/packages/dependency-extraction-webpack-plugin/README.md:

| Request | Global | Script handle |
| --- | --- | --- |
| `@babel/runtime/regenerator` | `regeneratorRuntime` | `wp-polyfill` |
| `@wordpress/*` | `wp['*']` | `wp-*` |
| `jquery` | `jQuery` | `jquery` |
| `lodash-es` | `lodash` | `lodash` |
| `lodash` | `lodash` | `lodash` |
| `moment` | `moment` | `moment` |
| `react-dom` | `ReactDOM` | `react-dom` |
| `react` | `React` | `react` |

Notably, `tinymce` support is dropped by this change.

sirreal committed:
Replace a8c webpack plugin with wp

sirreal commented on packages/calypso-build/webpack.config.js:
```diff
@@ -130,7 +130,7 @@ function getWebpackConfig(
 			} ),
 			new DuplicatePackageCheckerPlugin(),
-			...( env.WP ? [ new WordPressExternalDependenciesPlugin() ] : [] ),
+			...( env.WP ? [ new DependencyExtractionWebpackPlugin() ] : [] ),
```
Default behavior is probably best to include the polyfill. That's how `wp-scripts` build works.

```suggestion
			...( env.WP ? [ new DependencyExtractionWebpackPlugin( { injectPolyfill: true } ) ] : [] ),
```

sirreal committed:
Add polyfill by default

sirreal committed:
Update plugin for wpcom-block-editor

sirreal committed:
devDeps must live at the root

sirreal commented on apps/wpcom-block-editor/webpack.config.js:
```diff
@@ -57,6 +58,23 @@ function getWebpackConfig(
+			...webpackConfig.plugins.filter(
+				plugin => ! ( plugin instanceof DependencyExtractionWebpackPlugin )
+			),
+			new DependencyExtractionWebpackPlugin( {
```
This was added to match functionality from #32162 by @mmtr.

sirreal commented:
This seems to be working well now with the `process.env.FORCE_REDUCED_MOTION` changes. Reviews appreciated.

sirreal commented:
Collecting independent full e2e passing tests at 7883c35 because passing at the same time seems unlikely:
- [x] Mobile https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/20969
- [ ] Desktop …

ockham commented on package.json:
```diff
@@ -284,6 +284,7 @@
 		"@typescript-eslint/eslint-plugin": "1.10.2",
 		"@typescript-eslint/parser": "1.10.2",
 		"@wordpress/babel-plugin-makepot": "2.1.3",
+		"@wordpress/dependency-extraction-webpack-plugin": "^1.0.1",
```
Do we need/want it at the top level? Or should we move it to `apps/wpcom-block-editor/package.json`?

sirreal replied:
As a `devDependency`, it has to be at top level.

ockham commented:
Managed to get the desktop e2e tests to pass, https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/21002 :tada: 

Also did some smoke testing, specifically inserting a Classic block, and inserting an image into it (which opens the Calypso media modal). Works fine!

[One question](https://github.com/Automattic/wp-calypso/pull/34122/files#r311757651), otherwise good to :ship: !

ockham reviewed (approved):
(see https://github.com/Automattic/wp-calypso/pull/34122#issuecomment-519265600)

sirreal committed:
Update deps

sirreal commented:
Rebased to fix conflicts. I plan to merge when e2e passes again.

I'd appreciate if @Automattic/serenity could look to ensure the changes from #32162 have not regressed.

sirreal committed:
Compare against constructor.name for more resilience

We could get different versions of the same module, this should be more
resilient across versions.

sirreal commented:
> Rebased to fix conflicts. I plan to merge when e2e passes again.
> 
> I'd appreciate if @Automattic/serenity could look to ensure the changes from #32162 have not regressed.

I took a look and `tinymce/tinymce` does appear to be properly externed to global `tinymce`, but would appreciate confirmation 🙂

gwwar commented:
Do y'all have a matching server diff that we can test? See steps in PCYsg-l4k-p2

sirreal commented:
I built with `npx lerna run build --scope='@automattic/wpcom-block-editor'` and inspected the build output and contents. I'll get a diff ready for testing tomorrow 👍

sirreal commented:
> Do y'all have a matching server diff that we can test? See steps in [PCYsg-l4k-p2](https://href.li/?https://wp.me/PCYsg-l4k)

@gwwar I prepared D31387-code for testing.

gwwar commented:
Thanks @sirreal !
